riffdog/scanner.py

```diff
@@ -119,52 +132,113 @@ def _real_scan_element(scan_element, scanned, scanning):
+    except Exception as e:
+        # could be permission, bad file format, anything
+        logger.warn("Failed to read file - maybe not a state file for reason: %s" % e)
+
 
-def _s3_state_fetch(bucket_name):
+
+def _s3_inspector(state_location):
+    
+    logger.info("S3 State Inspecting %s" % state_location)
+    parts = state_location.split("/",1)
+    if len(parts) > 1:
+        if parts[1].endswith("/"):
+            # Fun Fact - Techincally keys can end in a / but lets assume the user isnt trying 
+            # to be awkward. - if anyone hits this, we will need to implement some kind of prefix
+            # marker so we know its a prefix or not.
+            _s3_state_fetch(parts[0], prefix=parts[1])
+        else:
+            _search_s3_state(parts[0], parts[1])
+    else:
+        _s3_state_fetch(state_location)
+
+
+def _s3_state_fetch(bucket_name, prefix=""):
     """
-    This function exists to fetch state files from S3
+    This function exists to locate state files from S3
     """
     client = boto3.client('s3')
-    s3 = boto3.resource('s3')
 
     logger.info("Processing bucket: %s" % bucket_name)
 
-    items = client.list_objects(Bucket=bucket_name, Prefix="")
+    items = client.list_objects(Bucket=bucket_name, Prefix=prefix)
 
     # _search_state(bucket_name, "sandbox_main/services/django_lambda_test/terraform.tfstate", s3)
 
-    for item in items.get('Contents', []):
-        if item['Key'].endswith("terraform.tfstate"): # FIXME: in future how can we better identify these files?
-            logging.info("Inspecting s3 item: %s" % item['Key'])
-            _search_state(bucket_name, item['Key'], s3)
+    for item in items.get('Contents', []):    
+        _search_s3_state(bucket_name, item['Key'])
 
 
-def _search_state(bucket_name, key, s3):
+def _search_s3_state(bucket_name, key):
+    logging.info("processing s3 file: %s" % key)
+
+    s3 = boto3.resource('s3')
+
     obj = s3.Object(bucket_name, key)
     content = obj.get()['Body'].read()
+
+    _search_state(key, content)
+
+
+def _search_state(filename, content):
+
     rd = ResourceDirectory()
     config = RDConfig()
     parsed = ""
     try:
         parsed = json.loads(content)
+        logger.info("Inspecting %s as a %s file" % (filename, parsed['version']))
 
         if parsed['version'] > 3:
             elements = parsed['resources']
         else:
-            elements = parsed['modules'][0]['resources'].values()
+            # elements = parsed['modules'][0]['resources'].values()
+            logger.warning("only state version v4 and above files are not supported right now")
+            # a bit of a cheeky return - not raising, just abandoning early.
+            return
 
         for res in elements:
-            if res['type'] in config.elements_to_scan:
+            if res['type'] in config.elements_to_scan or (res['type'] in rd.resource_aliases and rd.resource_aliases[res['type']] in config.elements_to_scan):
```

Comment:
I would probably split this line out so it could be unit tested

Reply:
So, this was actually a fix for a bug. I'm not sure how it would be split out without duplicating the if ? (i.e. for unit testing this is relatively easy with two tests - 

1) given a thing is in the elements_to_scan *or*
2) its an alias of something in elements_to_scan;

and then the negative tests of 'if a thing is not in elements to scan, nor an alias of something in there'.

Reply:
@jmons if we have those tests then I'm happy :)

Reply:
They will come ! they will come!